src/Riok.Mapperly/Descriptors/UnsafeAccessorContext.cs

```diff
@@ -0,0 +1,104 @@
+﻿using System.Globalization;
+using Microsoft.CodeAnalysis;
+using Riok.Mapperly.Descriptors.Mappings.MemberMappings.UnsafeAccess;
+using Riok.Mapperly.Helpers;
+
+namespace Riok.Mapperly.Descriptors;
+
+public class UnsafeAccessorContext
+{
+    private readonly UniqueNameBuilder _nameBuilder;
+    private readonly SymbolAccessor _symbolAccessor;
+    private readonly Dictionary<UnsafeAccessorKey, IUnsafeAccessor> _unsafeAccessors = new();
+
+    public UnsafeAccessorContext(UniqueNameBuilder nameBuilder, SymbolAccessor symbolAccessor)
+    {
+        _nameBuilder = nameBuilder.NewScope();
+        _symbolAccessor = symbolAccessor;
+    }
+
+    public IReadOnlyCollection<IUnsafeAccessor> UnsafeAccessors => _unsafeAccessors.Values;
+
+    public IUnsafeAccessor GetOrBuildAccessor(UnsafeAccessorType type, ISymbol symbol)
+    {
+        var key = new UnsafeAccessorKey(symbol, type);
+        if (_unsafeAccessors.TryGetValue(key, out var value))
+            return value;
+
+        var formatted = FormatAccessorName(symbol.Name);
+
+        var defaultMethodName = type switch
+        {
+            UnsafeAccessorType.GetField => $"Get{formatted}",
+            UnsafeAccessorType.GetProperty => $"Get{formatted}",
+            UnsafeAccessorType.SetProperty => $"Set{formatted}",
+            _ => throw new Exception($"Unknown {nameof(UnsafeAccessorType)}."),
+        };
+        var methodName = GetValidMethodName(symbol.ContainingType, defaultMethodName);
+
+        IUnsafeAccessor accessor = type switch
+        {
+            UnsafeAccessorType.GetField => new UnsafeFieldAccessor((IFieldSymbol)symbol, methodName),
+            UnsafeAccessorType.GetProperty => new UnsafeGetPropertyAccessor((IPropertySymbol)symbol, methodName),
+            UnsafeAccessorType.SetProperty => new UnsafeSetPropertyAccessor((IPropertySymbol)symbol, methodName),
+            _ => throw new Exception($"Unknown {nameof(UnsafeAccessorType)}."),
```

Comment:
Include the type, otherwise this might be quite hard to debug and probably use the more appropriate ArgumentOutOfRangeException here: throw new ArgumentOutOfRangeException(nameof(type), type, $"Unknown {nameof(UnsafeAccessorType)}.")